### PR TITLE
[Performance][SFA] Bypass redundant LightningIndexer scoring for eligible full-visible short-prefill cases

### PR DESCRIPTION
**Pull request by Ruiqiu-Zheng**
### What this PR does / why we need it?

This PR rebases the opt-in full-visible short-prefill LightningIndexer bypass onto current main after the indexer/SFA ownership split in #15669.

When the entire visible KV window fits within `topk=2048`, LightningIndexer scoring cannot remove any valid KV entry. For this bounded case, `AscendSFAIndexerBackend.forward()` returns a slice of a precomputed round-robin full-visible index table after the normal indexer cache write, while every unsupported or unproven case falls back to the incumbent LightningIndexer path. The rebased patch does not put bypass ownership back into `attention/sfa_v1.py`.

The feature is disabled by default through `enable_sfa_full_visible_index_bypass=False`. Current admission is deliberately fail-closed: single sequence, supported prefill mode, total visible KV `<=2048`, block size 128, supported model/layer/device/table geometry. MTP/speculative paths, LI C8, PCP/DSA-CP, missing or mismatched metadata/device/table state, visible KV `>2048`, and other unproven cells keep the native path.

The shared RR table is device-keyed and read-only by contract. The worker establishes the NPU device before model/model-runner construction; request-time device mismatch declines the fast path rather than allocating a replacement table. Downstream index-cache publication copies into a separate buffer, so the shared table is not the mutation destination.

### Does this PR introduce any user-facing change?

Yes. It documents the new opt-in `enable_sfa_full_visible_index_bypass` additional configuration option. The default is `False`, so existing behavior is unchanged unless explicitly enabled.

### Validation

The semantic-rebased source is frozen against the R52 candidate. Final-source host hygiene passed (`ruff format --check`, Ruff E/F, `py_compile`, and `git diff --check`).

Single-NPU native validation on the rebased source covers full-visible cases `(context, query) = (0,64), (128,64), (0,2048)`. The bypass and native LightningIndexer select the same valid KV set for every row; downstream native SFA outputs are equivalent within the recorded tolerance; repeated calls are exact; the shared table remains unchanged; and `>2048` falls back. Native LI is allowed to return a different ordering for the same full-visible selected set, so the correctness oracle is selected-set equality plus downstream SFA equivalence rather than elementwise index-order equality.

Bounded single-NPU LI+SFA path measurements on the validated path previously showed approximately 10.75% (`K=64`), 8.44% (`K=1024`), and 6.02% (`K=2048`) latency reduction, with all measured blocks positive. These are component/path measurements only, not full-model TTFT or production-performance claims.

### Scope / non-claims

This PR does not claim full-model or production improvement, and it does not change MTP/speculative, LI C8, PCP, DSA-CP, or long-context behavior. Unsupported cases remain on the incumbent native path.

- vLLM main: https://github.com/vllm-project/vllm/commit/a97dacb7106ee49f39f3d1fc6ae1800ff724e01d

### CODE REVIEW

File: vllm_ascend/attention/sfa_v1.py [hunk 1]
```diff
@@ -530,6 +535,61 @@ def __init__(
...
+            or attn_metadata.num_actual_tokens != topk_num_tokens
+            or self._full_visible_index_table is None
+            or self._full_visible_index_table.device != device
+        ):
+            return None
+        kv_length = int(attn_metadata.seq_lens_cpu[0])
+        context_length = kv_length - attn_metadata.num_actual_tokens
+        if context_length < 0 or kv_length > SFA_INDEXER_SPARSE_COUNT:
+            return None
+        return self._full_visible_index_table[context_length + 1 : kv_length + 1].unsqueeze(1)
```
Comment (nu11ptr7) [inline vllm_ascend/attention/sfa_v1.py:594]: This returns a row-slice *view* of a class-level persistent table shared by every layer (and never reset per request). If anything downstream writes into `topk_indices` in place - e.g. `_update_indexcache_topk_indices` or a kernel - it would corrupt other layers'/requests' index rows. Worth confirming all consumers are read-only, or returning a clone.

Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py:594]: Thanks for pointing out this potential aliasing issue.

We traced the consumers of the full-visible index tensor. In the inspected bypass path, `_update_indexcache_topk_indices` uses `copy_` into the separately allocated `topk_indices_buffer`; `squeeze` only changes the source view. The persistent full-visible table is not the cache-update destination.

We verified whole-table invariance with separate host and native canaries at PR commit `1eaa5d2ff765fa80551536f1990560ae93caed63`. The native canary covered 8 cases and 28 synchronized SFA consumer calls, including overlapping short/longer/short slices, two extracted backend objects, and cache update before consumption. The entire 2049 × 2048 table remained exactly equal to an independently allocated pre-consumption snapshot.

These results cover the extracted helper/cache/adapter path with the installed native SFA extension on Ascend910B2 in eager BF16 mode; they do not establish full-model, graph, or future-consumer safety. Based on this bounded evidence, we keep the zero-copy view instead of adding an unconditional clone.

Thanks for the careful review.

[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py]: Thanks for checking device placement. This table is not a CPU tensor that relies on a later `Module.to()` call: the backend constructor explicitly selects `torch.device("npu", torch.npu.current_device())` and copies the table to that device. The cache is keyed by device. The worker initialization also explicitly selects its NPU before creating the model runner.

For this bounded fast path, a missing table or device mismatch deliberately falls back to the existing LightningIndexer path. Creating a table in this helper would introduce table construction and a device transfer into the request path, which this PR intentionally keeps as a slice/view lookup.

We therefore retain the initialization-time allocation and fallback checks. Arbitrary device migration after backend construction is not accelerated by this path; a mismatch disables the optimization rather than returning indices on the wrong device. This is a source-based explanation, not a new cross-device runtime test claim.

Sources: sfa_v1.py at 213451e04c560ca32546058b83b0ae09070d8638 lines 538–594; worker/worker.py lines 421–497.

File: tests/ut/attention/test_sfa_v1.py [hunk 2]
```diff
@@ -886,6 +888,156 @@ def _setup_kv_b_proj(self):
...
+
+        metadata.block_size = SFA_FULL_VISIBLE_TEMPLATE_BLOCK_SIZE
+        self.assertIsNone(impl._get_full_visible_topk_indices(metadata, 64, torch.device("meta")))
+
+        metadata.seq_lens_cpu = None
+        self.assertIsNone(impl._get_full_visible_topk_indices(metadata, 64, torch.device("cpu")))
+
+        metadata.seq_lens_cpu = torch.tensor([64])
+        impl._full_visible_index_table = None
+        self.assertIsNone(impl._get_full_visible_topk_indices(metadata, 64, torch.device("cpu")))
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply tests/ut/attention/test_sfa_v1.py]: This assertion intentionally protects the fallback contract described in the implementation thread: if the persistent table is missing, the helper returns `None` and the caller uses the existing LightningIndexer path. Changing the assertion to require lazy initialization would change that contract and add allocation to the request path, so we are keeping the fallback test.

File: vllm_ascend/ascend_config.py [hunk 3]
```diff
@@ -414,6 +414,7 @@ class AscendConfig:
     # ---- A-family (envs fallback): default = envs module value, before-validator injects ----
     enable_fused_mc2: int = 0
     enable_mlapo: bool = True
+    enable_sfa_full_visible_index_bypass: bool = False
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/ascend_config.py:420]: Agreed — this is a user-facing opt-in and the current configuration reference is missing an entry.

The entry needs to cover the default `False`, an `--additional-config` example, the single-sequence prefill restriction, the 2,048-token limit on cached context plus new tokens, normal-layer permission, decode/MTP/C8 and other unsupported-path fallbacks, and the persistent per-process/device table allocation (16,785,408 bytes, approximately 16.008 MiB), shared across layers rather than allocated per request.

This remains a documentation follow-up at the current PR HEAD; it does not require changing the runtime algorithm.

[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/ascend_config.py:420]: Addressed in the current rebased head `180baa0795cba9801a9a9953017f5fd3e1ac1e70`. `docs/source/user_guide/configuration/additional_config.md` now documents `enable_sfa_full_visible_index_bypass`, including the default-off opt-in, single-sequence prefill eligibility, total visible KV <= 2048 (including cached context), MTP/C8/CP/unsupported-mode fallbacks, and the shared per-process/device table memory cost. The Markdown link and Read the Docs checks pass on this head.

**Comment by Ruiqiu-Zheng** [comment]
The earlier merge conflict is now resolved. This PR is rebased directly on current `main` `e5118d151314ae18e56c0b63aa8dd00d294adc22`; current HEAD is `180baa0795cba9801a9a9953017f5fd3e1ac1e70` and GitHub reports it mergeable. On this head, DCO, `pre-commit`, CPU UT, Markdown link check, and both documentation builds pass. The requested additional-config documentation is also included.

The remaining `ci-gate` failure is only the ready-label gate; selected tests have not run yet. Could a maintainer please add `ready-precise` to run the focused selected tests?
